File: pipeline2/lib/tracker.go
package lib

/*
for 1 fps

* threshold means we look at more frames when confidence is less than threshold
* threshold=0: don't look at any more frames
* threshold=1: don't track at this rate (just start tracking at next framerate)

pre-process:
(1) track over the video at 1 fps; but update RNN features using ground truth tracks
(2) for each track, it yields:
	(a) 1 - (2nd highest) / (1st highest): so if threshold is higher, then we would look at extra frame
	(b) minimum threshold needed to get it correct, i.e., 0 if correct option had highest prob, or 1 - (correct score) / (1st highest) otherwise
(3) using (b), compute minimum threshold needed to recover 80% contiguous segment of each track, call this (c)

setting threshold:
(1) iterate over alpha [0.99, 0.98, 0.96, 0.92, ...]
(2) pick alpha-percentile value in the (c) values
(3) estimate how many extra frames we would use based on the (a) values
*/

import (
	"bufio"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"sync"
)

type Tracker struct {
	cmd *exec.Cmd
	stdin io.WriteCloser
	rd *bufio.Reader
	mu sync.Mutex
}

type TrackerPacket struct {
	ID int `json:"id"`
	FrameIdx int `json:"frame_idx"`
	Detections []Detection `json:"detections"`
	GT map[int]*int `json:"gt"`
	Type string `json:"type"`
}

type TrackerResponse struct {
	Outputs []int `json:"outputs"`
	Conf float64 `json:"conf"`
	T map[int]float64 `json:"t"`
}

func NewTracker(dataRoot string, mode string, cfg Config) *Tracker {
	var pythonFname string
	if mode == "iou" {
		pythonFname = "tracker.py"
	} else if mode == "rnn" {
		pythonFname = "../rnn/tracker.py"
	} else if mode == "miris" {
		pythonFname = "../tracker-miris/tracker.py"
	} else {
		panic(fmt.Errorf("bad tracker mode %s", mode))
	}
	cmd := exec.Command(
		"python", pythonFname,
		dataRoot, cfg.Label,
		strconv.Itoa(cfg.OrigDims[0]), strconv.Itoa(cfg.OrigDims[1]),
	)
	stdin, err := cmd.StdinPipe()
	if err != nil {
		panic(err)
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		panic(err)
	}
	if true {
		cmd.Stderr = os.Stderr
	}
	if err := cmd.Start(); err != nil {
		panic(err)
	}
	rd := bufio.NewReader(stdout)
	return &Tracker{
		cmd: cmd,
		stdin: stdin,
		rd: rd,
	}
}

func (tracker *Tracker) do(id int, frameIdx int, im Image, detections []Detection, gt map[int]*int) TrackerResponse {
	packet := TrackerPacket{
		ID: id,
		FrameIdx: frameIdx,
		Detections: detections,
		GT: gt,
		Type: "job",
	}
	bytes := JsonMarshal(packet)

	tracker.mu.Lock()
	defer tracker.mu.Unlock()

	header := make([]byte, 4)
	binary.BigEndian.PutUint32(header, uint32(len(bytes)))
	tracker.stdin.Write(header)
	tracker.stdin.Write(bytes)

	header = make([]byte, 12)
	binary.BigEndian.PutUint32(header[0:4], uint32(len(im.Bytes)))
	binary.BigEndian.PutUint32(header[4:8], uint32(im.Width))
	binary.BigEndian.PutUint32(header[8:12], uint32(im.Height))
	tracker.stdin.Write(header)
	tracker.stdin.Write(im.Bytes)

	var line string
	for {
		var err error
		line, err = tracker.rd.ReadString('\n')
		if err != nil {
			panic(err)
		}
		line = strings.TrimSpace(line)
		if !strings.HasPrefix(line, "json") {
			continue
		}
		break
	}

	jsonBytes := []byte(line[4:])
	var response TrackerResponse
	if err := json.Unmarshal(jsonBytes, &response); err != nil {
		panic(err)
	}
	return response
}

func (tracker *Tracker) Infer(id int, frameIdx int, im Image, detections []Detection) ([]Detection, float64) {
	response := tracker.do(id, frameIdx, im, detections, nil)

	var outputs []Detection
	for i, d := range detections {
		d.TrackID = new(int)
		*d.TrackID = response.Outputs[i]
		outputs = append(outputs, d)
	}
	return outputs, response.Conf
}

func (tracker *Tracker) End(id int) {
	packet := TrackerPacket{
		ID: id,
		Type: "end",
	}
	bytes := JsonMarshal(packet)

	tracker.mu.Lock()
	header := make([]byte, 4)
	binary.BigEndian.PutUint32(header, uint32(len(bytes)))
	tracker.stdin.Write(header)
	tracker.stdin.Write(bytes)
	tracker.mu.Unlock()
}

type TrackerProfile struct {
	NumFrames int
	Thresholds []float64
}

func (prof TrackerProfile) MaxGap() int {
	if len(prof.Thresholds) < 2 {
		return 1
	}
	gap := 1
	for _, threshold := range prof.Thresholds[1:] {
		if threshold == 1 {
			return gap
		}
		gap *= 2
	}
	return gap
}

// Returns map from alpha value to tracker profile
func (tracker *Tracker) GetThresholds(videoPath string, jsonPath string, origDims [2]int, detectorDims [2]int, freqs []int) map[float64]TrackerProfile {
	// returns length of longest contiguous, non-disconnected segment
	getLongestSegment := func(disconnects []bool) int {
		discList := []int{-1}
		for i, bad := range disconnects {
			if !bad {
				continue
			}
			discList = append(discList, i)
		}
		discList = append(discList, len(disconnects)-1)

		longest := 0
		for i := 0; i < len(discList)-1; i++ {
			l := discList[i+1]-discList[i]
			if l > longest {
				longest = l
			}
		}
		return longest
	}

	// get the minimum threshold needed to recover this track
	getMinThreshold := func(confs []float64) float64 {
		// iteratively disconnect the track at edges with smallest confidence
		// repeat until longest contiguous segment is <80% of the original track
		disconnects := make([]bool, len(confs))
		var threshold float64
		for {
			// find highest confidence
			var bestConf float64 = 0.01
			var bestIdx int = -1
			for i, score := range confs {
				if disconnects[i] {
					continue
				}
				if score > bestConf {
					bestIdx = i
					bestConf = score
				}
			}
			if bestIdx == -1 {
				break
			}

			// disconnect and check longest segment length
			disconnects[bestIdx] = true
			l := getLongestSegment(disconnects)
			if l > len(disconnects)*8/10 {
				continue
			}
			// it doesn't work, so threshold must be at least as high as bestConf
			threshold = bestConf
			break
		}
		return threshold
	}

	// determine whether inferred detections match up for a given gt track
	isMatchingTrack := func(gtID int, freq int, gt [][]Detection, detections [][]Detection) bool {
		// find the track ID in detections that matches most with gtID
		hits := make(map[int]int)
		for frameIdx := range gt {
			if frameIdx%freq != 0 {
				continue
			}
			for i, d := range gt[frameIdx] {
				if *d.TrackID != gtID {
					continue
				}
				hits[*detections[frameIdx][i].TrackID]++
			}
		}
		var bestID int = -1
		for trackID := range hits {
			if hits[trackID] < 2 {
				continue
			}
			if bestID == -1 || hits[trackID] > hits[bestID] {
				bestID = trackID
			}
		}
		if bestID == -1 {
			return false
		}

		// see if inferred track matches at 80% of the detections
		var matches int = 0
		var misses int = 0
		for frameIdx := range gt {
			if frameIdx%freq != 0 {
				continue
			}
			for i, d := range gt[frameIdx] {
				if *d.TrackID != gtID {
					continue
				}
				if *detections[frameIdx][i].TrackID == bestID {
					matches++
				} else {
					misses++
				}
			}
		}

		return matches >= 4*misses
	}

	getConfidences := func(id int, freq int) (trackMinThreshs []float64, frameConfidences []float64) {
		var detections [][]Detection
		ReadJsonFile(jsonPath+fmt.Sprintf("%d.json", id), &detections)

		// re-scale detections and compute the range of frames each track spans
		//RescaleDetections(detections, origDims, detectorDims)
		trackRanges := make(map[int][2]int)
		for i := range detections {
			for j := range detections[i] {
				trackID := *detections[i][j].TrackID
				if _, ok := trackRanges[trackID]; !ok {
					trackRanges[trackID] = [2]int{i, i}
				} else {
					trackRanges[trackID] = [2]int{trackRanges[trackID][0], i}
				}
			}
		}
		activeTracks := make([][]int, len(detections))
		for trackID, rng := range trackRanges {
			if rng[1]-rng[0] < 96 {
				// this track is rather noisy so we shouldn't use it in scoring
				continue
			}
			for frameIdx := rng[0]; frameIdx <= rng[1]; frameIdx++ {
				activeTracks[frameIdx] = append(activeTracks[frameIdx], trackID)
			}
		}

		// run the tracker to get probabilities but force it to update using gt tracks
		// although we also separately collect the inferred tracks without gt
		confByTrackID := make(map[int][]float64)
		inferred := make([][]Detection, len(detections))
		vreader := ReadFfmpeg(filepath.Join(videoPath, fmt.Sprintf("%d.mp4", id)), detectorDims[0], detectorDims[1])
		for frameIdx := 0; ; frameIdx++ {
			im, err := vreader.Read()
			if err == io.EOF {
				break
			} else if err != nil {
				panic(err)
			}

			if frameIdx%freq != 0 {
				continue
			}

			gt := make(map[int]*int)
			for _, trackID := range activeTracks[frameIdx] {
				gt[trackID] = nil
			}
			for i, d := range detections[frameIdx] {
				if _, ok := gt[*d.TrackID]; !ok {
					// track was too short
					continue
				}
				gt[*d.TrackID] = new(int)
				*gt[*d.TrackID] = i
			}

			response := tracker.do(id, frameIdx, im, detections[frameIdx], gt)
			frameConfidences = append(frameConfidences, response.Conf)
			for trackID, thresh := range response.T {
				confByTrackID[trackID] = append(confByTrackID[trackID], thresh)
			}

			// also get the inferred tracks
			inferred[frameIdx], _ = tracker.Infer(-1, frameIdx, im, detections[frameIdx])
		}
		tracker.End(id)
		tracker.End(-1)
		vreader.Close()

		// get min thresholds for the tracks
		// if inferred track overlaps >= 80% with gt track, then we mark it okay
		// otherwise threshold is based on the probabilities when running it with gt updates
		for gtID, confs := range confByTrackID {
			var minThreshold float64
			if isMatchingTrack(gtID, freq, detections, inferred) {
				minThreshold = 0
			} else {
				minThreshold = getMinThreshold(confs)
			}
			trackMinThreshs = append(trackMinThreshs, minThreshold)
		}

		bytes := JsonMarshal(inferred)
		ioutil.WriteFile("/tmp/x.json", bytes, 0644)

		return
	}

	getWorkload := func(frameConfidences []float64, threshold float64) int {
		var numFrames int
		for _, conf := range frameConfidences {
			if conf > threshold {
				continue
			}
			numFrames++
		}
		return numFrames
	}

	// list videos
	var ids []int
	files, err := ioutil.ReadDir(videoPath)
	if err != nil {
		panic(err)
	}
	for _, fi := range files {
		if !strings.HasSuffix(fi.Name(), ".mp4") {
			continue
		}
		id := ParseInt(strings.Split(fi.Name(), ".mp4")[0])
		ids = append(ids, id)
	}

	// first get min thresholds and frame confidences at each freq
	minThresholds := make([][]float64, len(freqs))
	frameConfidences := make([][]float64, len(freqs))
	for i, freq := range freqs {
		for _, id := range ids {
			fmt.Printf("process %d.mp4 at freq=%d\n", id, freq)
			curThreshs, curConfidences := getConfidences(id, freq)
			minThresholds[i] = append(minThresholds[i], curThreshs...)
			frameConfidences[i] = append(frameConfidences[i], curConfidences...)
		}
		sort.Float64s(minThresholds[i])
	}

	// now, profile each alpha
	profiles := make(map[float64]TrackerProfile)
	for _, alpha := range []float64{1.0, 0.998, 0.995, 0.99, 0.98, 0.96, 0.92, 0.84, 0.68} {
		var profile TrackerProfile
		for i := range freqs {
			// get alpha-percentile lenient (highest) threshold
			idx := int(float64(len(minThresholds[i]))*alpha)
			if idx >= len(minThresholds[i]) {
				idx = len(minThresholds[i])-1
			}
			threshold := minThresholds[i][idx]

			profile.Thresholds = append(profile.Thresholds, threshold)
			profile.NumFrames += getWorkload(frameConfidences[i], threshold)
		}
		profiles[alpha] = profile
	}

	return profiles
}

func (tracker *Tracker) Close() {
	tracker.stdin.Close()
	tracker.cmd.Wait()
}

File: pipeline2/tracker.py
from discoverlib import geom

import json
import numpy
import struct
import sys

def eprint(s):
	sys.stderr.write(s+"\n")
	sys.stderr.flush()

class Detection(object):
	def __init__(self, d, frame_idx):
		self.frame_idx = frame_idx
		self.left = d['left']
		self.top = d['top']
		self.right = d['right']
		self.bottom = d['bottom']
		self.cls = d['class']

class Tracker(object):
	def __init__(self):
		# for each active object, a list of detections
		self.objects = {}
		# most recent frame we processed
		self.last_frame = -1
		# object id counter
		self.next_id = 0

	# returns:
	# (1) track IDs for detections in the current frame
	# (2) the confidence of this frame, i.e., min(1 - 2nd highest score / highest score) taken over active objects
	# (3) if gt set: map from active objects to the minimum threshold needed for this frame
	#	 - if highest prob matched gt: min threshold is 0
	#	 - else: threshold is 1 - (gt score) / (highest score)
	# gt is a map from active object id -> the idx of current frame detections that object should match with
	# if an object id doesn't appear in gt, then we remove it from active set
	def update(self, frame_idx, im, detections, gt=None):
		# cleanup frames that are now in future if needed
		if frame_idx < self.last_frame:
			for id in list(self.objects.keys()):
				self.objects[id] = [d for d in self.objects[id] if d.frame_idx < frame_idx]
				if len(self.objects[id]) == 0:
					del self.objects[id]
		self.last_frame = frame_idx

		# helper func to estimate speed of an object
		def estimate_speed(dlist):
			if len(dlist) == 1:
				return (0, 0)
			last = dlist[-1]
			n = max(5, frame_idx - last.frame_idx)
			target_frame = last.frame_idx - n
			best = None
			for d in dlist[:-1]:
				if best is None or abs(d.frame_idx-target_frame) < abs(best.frame_idx-target_frame):
					best = d
			dx = (last.left-best.left+last.right-best.right)/2.0
			dy = (last.top-best.top+last.bottom-best.bottom)/2.0
			scale = float(frame_idx-last.frame_idx)/float(last.frame_idx-best.frame_idx)
			return (dx*scale, dy*scale)

		# helper func to estimate position of an object at current frame
		def estimate_pos(dlist):
			dx, dy = estimate_speed(dlist)
			last = dlist[-1]
			return geom.Rectangle(
				geom.Point(last.left+dx, last.top+dy),
				geom.Point(last.right+dx, last.bottom+dy),
			)

		# helper func to extract matches from a matrix
		def extract_matches(mat):
			votes = {}
			for i in range(mat.shape[0]):
				for j in range(mat.shape[1]):
					p = mat[i, j]
					if i not in votes or p > votes[i][0]:
						if j == mat.shape[1]-1:
							votes[i] = (p, None)
						else:
							votes[i] = (p, j)
			# group by receiver and vote on max idx2->idx1 to eliminate duplicates
			votes2 = {}
			for idx1, t in votes.items():
				p, idx2 = t
				if idx2 is not None and (idx2 not in votes2 or p > votes2[idx2][0]):
					votes2[idx2] = (p, idx1)
			return {idx1: idx2 for (idx2, (_, idx1)) in votes2.items()}

		# match each object with current frame
		objects = list(self.objects.items())
		mat = numpy.zeros((len(objects), len(detections)+1), dtype='float32')
		id_to_obj_idx = {}
		obj_idx_to_id = {}
		for i, (id, dlist) in enumerate(objects):
			id_to_obj_idx[id] = i
			obj_idx_to_id[i] = id
			r1 = estimate_pos(dlist)
			for j, d2 in enumerate(detections):
				r2 = geom.Rectangle(
					geom.Point(d2['left'], d2['top']),
					geom.Point(d2['right'], d2['bottom']),
				)
				iou = r1.iou(r2)
				if d2['class'] != dlist[-1].cls:
					iou = 0
				mat[i, j] = iou
			mat[i, len(detections)] = 0.1

		# update objects based either on mat or gt
		min_thresholds = {}
		if gt is None:
			matches = extract_matches(mat)
			track_ids = [None]*len(detections)
			for idx1, idx2 in matches.items():
				d = detections[idx2]
				objects[idx1][1].append(Detection(d, frame_idx))
				track_ids[idx2] = obj_idx_to_id[idx1]
			for idx2, d in enumerate(detections):
				if track_ids[idx2] is not None:
					continue
				id = self.next_id
				self.next_id += 1
				self.objects[id] = [Detection(d, frame_idx)]
				track_ids[idx2] = id

			# remove old objects
			for id in list(self.objects.keys()):
				age = frame_idx - self.objects[id][-1].frame_idx
				if age < 20:
					continue
				del self.objects[id]
		else:
			# remove objects not in gt
			for id in list(self.objects.keys()):
				if id not in gt:
					del self.objects[id]

			# update objects
			track_ids = [-1]*len(detections)
			for id, idx2 in gt.items():
				if idx2 is None:
					min_thresholds[id] = 0.0
					continue

				d = detections[idx2]
				if id not in self.objects:
					self.objects[id] = [Detection(d, frame_idx)]
				else:
					self.objects[id].append(Detection(d, frame_idx))
				track_ids[idx2] = id

				# also compute per-object min thresholds here
				if id not in id_to_obj_idx:
					min_thresholds[id] = 0.0
					continue
				idx1 = id_to_obj_idx[id]
				if numpy.argmax(mat[idx1, :]) == idx2:
					min_thresholds[id] = 0.0
					continue
				min_thresholds[id] = float(1 - (mat[idx1, idx2]+0.01)/(mat[idx1, :].max()+0.01))

		# compute frame confidence
		if mat.shape[0] > 0:
			high1 = mat.max(axis=1)
			mat_ = numpy.copy(mat)
			mat_[numpy.arange(mat_.shape[0]), numpy.argmax(mat_, axis=1)] = 0
			high2 = mat_.max(axis=1)
			track_confidences = 1 - (high2+0.01)/(high1+0.01)
			frame_confidence = float(numpy.min(track_confidences))
		else:
			frame_confidence = 1.0

		return track_ids, frame_confidence, min_thresholds

if __name__ == '__main__':
	stdin = sys.stdin.detach()
	trackers = {}

	def read_im():
		buf = stdin.read(12)
		if not buf:
			return None
		(l, width, height) = struct.unpack('>III', buf)
		buf = stdin.read(l)
		return numpy.frombuffer(buf, dtype='uint8').reshape((height, width, 3))

	def read_json():
		buf = stdin.read(4)
		if not buf:
			return None
		(l,) = struct.unpack('>I', buf)
		buf = stdin.read(l)
		return json.loads(buf.decode('utf-8'))

	while True:
		packet = read_json()
		if packet is None:
			break
		id = packet['id']
		msg = packet['type']

		if msg == 'end':
			del trackers[id]
			continue

		im = read_im()

		detections = packet['detections']
		if detections is None:
			detections = []

		gt = None
		if 'gt' in packet and packet['gt']:
			gt = {int(id): idx for id, idx in packet['gt'].items()}

		if id not in trackers:
			trackers[id] = Tracker()
		track_ids, frame_confidence, min_thresholds = trackers[id].update(packet['frame_idx'], im, detections, gt=gt)
		d = {
			'outputs': track_ids,
			'conf': frame_confidence,
			't': min_thresholds,
		}
		sys.stdout.write('json'+json.dumps(d)+'\n')
		sys.stdout.flush()
